### PR TITLE
[PyTorch] Add fused kernel for binary_cross_entropy_with_logits

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #78269
* #78268
* __->__ #78267
* #78266
* #78265

I noticed this does a ton of extra dispatches (and, on CUDA, kernel launches), so I'm adding fused kernels.

Differential Revision: [D36650625](https://our.internmc.facebook.com/intern/diff/D36650625/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D36650625/)!

### CODE REVIEW

**Review by jbschlosser — COMMENTED** [review]
Got any benchmark results?

File: aten/src/ATen/native/cuda/Loss.cu [hunk 1]
```diff
@@ -144,6 +144,84 @@ Tensor& binary_cross_entropy_out_cuda(const Tensor& input, const Tensor& target,
...
+  }
+
+  if (reduction != at::Reduction::None) {
+    Tensor loss_reduced;
+    if (reduction == at::Reduction::Mean) {
+      loss_reduced = loss.mean();
+    } else if (reduction == at::Reduction::Sum) {
+      loss_reduced = loss.sum();
+    }
+    loss.resize_as_(loss_reduced).copy_(loss_reduced);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Loss.cu:207]: looks like you should be able to avoid this copy? You are not obligated to have `*_out_cuda` function that has to get real `loss` argument that you can modify only inplace, so you can just return `loss_reduced`.

**Comment by swolchok** [comment]
> Got any benchmark results?

no. what benchmark?

**Comment by jbschlosser** [comment]
> no. what benchmark?

Just curious about general perf on CPU and CUDA before and after the fused kernel.
